expertsystem/state/particle.py

```diff
@@ -258,6 +264,198 @@ def convert_to_dict(
+            for sub_list in nested_list
+        ]
+
+    @staticmethod
+    def __sort(nested_list: List[List[Any]]) -> List[List[Any]]:
+        return sorted([sorted(sub_list) for sub_list in nested_list])
+
+    @staticmethod
+    def __extract_particle_name(item: object) -> str:
+        if isinstance(item, str):
+            return item
+        if isinstance(item, (tuple, list)) and isinstance(item[0], str):
+            return item[0]
+        if isinstance(item, Particle):
+            return item.name
+        if isinstance(item, dict) and "Name" in item:
+            return str(item["Name"])
+        raise ValueError(
+            f"Cannot extract particle name from {item.__class__.__name__}"
+        )
+
+
+def get_kinematic_representation(
+    graph: StateTransitionGraph[StateWithSpins],
+) -> KinematicRepresentation:
+    r"""Group final or initial states by node, sorted by length of the group.
+
+    The resulting sorted groups can be used to check whether two
+    `.StateTransitionGraph` instances are kinematically identical. For
+    instance, the following two graphs:
+
+    .. code-block::
+
+        J/psi -- 0 -- pi0
+                  \
+                   1 -- gamma
+                    \
+                     2 -- gamma
+                      \
+                       pi0
+
+        J/psi -- 0 -- pi0
+                  \
+                   1 -- gamma
+                    \
+                     2 -- pi0
+                      \
+                       gamma
+
+    both result in:
+
+    .. code-block::
+
+        kinematic_representation.final_state == \
+            [["gamma", "gamma"], ["gamma", "gamma", "pi0"], \
+             ["gamma", "gamma", "pi0", "pi0"]]
+        kinematic_representation.initial_state == \
+            [["J/psi"], ["J/psi"]]
+
+    and are therefore kinematically identical. The nested lists are sorted (by
+    `list` length and element content) for comparisons.
+
+    Note: more precisely, the states represented here by a `str` only also have
+    a list of allowed spin projections, for instance, :code:`("J/psi", [-1,
+    +1])`. Note that a `tuple` is also sortable.
+    """
+
+    def get_state_groupings(
+        edge_per_node_getter: Callable[[int], List[int]]
+    ) -> List[List[int]]:
+        return [edge_per_node_getter(i) for i in graph.nodes]
+
+    def fill_groupings(grouping_with_ids: List[List[Any]]) -> List[List[Any]]:
+        return [
+            [graph.edge_props[edge_id] for edge_id in group]
+            for group in grouping_with_ids
+        ]
+
+    initial_state_edge_groups = fill_groupings(
+        get_state_groupings(graph.get_originating_initial_state_edges)
+    )
+    final_state_edge_groups = fill_groupings(
+        get_state_groupings(graph.get_originating_final_state_edges)
+    )
+    return KinematicRepresentation(
+        initial_state=initial_state_edge_groups,
+        final_state=final_state_edge_groups,
+    )
```

Comment:
I would make this the constructor of the `KinematicalRepresenation` (the docstring can be merged). That is the only use case that is needed and it will not let that cryptic underlying representation (triple nested lists) shine through the class.

Reply:
This constructor is currently used here:
https://github.com/ComPWA/expertsystem/blob/8b0660eda9d2acc817682dab3c1c5a648d48c455/expertsystem/state/particle.py#L678-L681

Reply:
In other words, `__init__` is required to convert the `allowed_final_state_groupings` that the user provides into a `KinematicRepresentation` while `get_kinematic_representation` is required to compare the `StateTransitionGraph` to those `allowed_final_state_groupings`.

Reply:
What could be done is to turn it into a method `from_graph`, like this method:
https://github.com/ComPWA/expertsystem/blob/03725998271da654aace08f109b07d12ee05dcbf/expertsystem/topology.py#L349-L352

Reply:
All right. Then lets leave it as it is for now. For the future it would be good to get this merged somehow